### PR TITLE
expression: implement vectorized evaluation for `builtinLocate3ArgsSig`

### PR DESCRIPTION
**Pull request by b41sh**
### What problem does this PR solve? 
Implement vectorized evaluation for `builtinLocate3ArgsSig`.
[12106](https://github.com/pingcap/tidb/issues/12106)

### What is changed and how it works?
```
goos: darwin
goarch: amd64
pkg: github.com/pingcap/tidb/expression
BenchmarkVectorizedBuiltinStringFunc/builtinLocate3ArgsSig-VecBuiltinFunc-4     5688  213284 ns/op  13488 B/op  932 allocs/op
BenchmarkVectorizedBuiltinStringFunc/builtinLocate3ArgsSig-NonVecBuiltinFunc-4  4701  258709 ns/op  13488 B/op  932 allocs/op
```

### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_string_vec.go [hunk 1]
```diff
@@ -333,11 +333,69 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
...
+	defer b.bufAllocator.put(buf2)
+	if err := b.args[2].VecEvalInt(b.ctx, input, buf2); err != nil {
+		return err
+	}
+
+	result.ResizeInt64(n, false)
+	i64s := result.Int64s()
+	poses := buf2.Int64s()
+	for i := 0; i < n; i++ {
+		if buf.IsNull(i) || buf1.IsNull(i) || buf2.IsNull(i) {
```
Comment (qw4990) [inline expression/builtin_string_vec.go]: We can use `result.MergeNulls(buf, buf1, buf2)`.

File: expression/builtin_string_vec.go [hunk 2]
```diff
@@ -333,11 +333,69 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
...
+	}
+	buf1, err := b.bufAllocator.get(types.ETString, n)
+	if err != nil {
+		return err
+	}
+	defer b.bufAllocator.put(buf1)
+	if err := b.args[1].VecEvalString(b.ctx, input, buf1); err != nil {
+		return err
+	}
+	buf2, err := b.bufAllocator.get(types.ETInt, n)
```
Comment (qw4990) [inline expression/builtin_string_vec.go]: How about using `result` as `buf2` here?

Comment (b41sh) [reply expression/builtin_string_vec.go]: goot idea

File: expression/builtin_string_vec_test.go [hunk 3]
```diff
@@ -29,8 +29,14 @@ var vecBuiltinStringCases = map[string][]vecExprBenchCase{
 	ast.Convert:        {},
 	ast.Substring:      {},
 	ast.SubstringIndex: {},
-	ast.Locate:         {},
-	ast.Hex:            {},
+	ast.Locate: {
+		{
+			retEvalType:   types.ETInt,
+			childrenTypes: []types.EvalType{types.ETString, types.ETString, types.ETInt},
+			geners:        []dataGenerator{&randLenStrGener{0, 10}, &randLenStrGener{0, 20}, &rangeInt64Gener{-10, 20}},
```
Comment (qw4990) [inline expression/builtin_string_vec_test.go]: How about adding one more case where the length of `substrs` is between `1` to `2` to increase the probability that the `substr` is included in the `str`.

File: expression/builtin_string_vec.go [hunk 4]
```diff
@@ -333,11 +333,63 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
...
+		subStrLen := len([]rune(subStr))
+		if pos < 0 || pos > int64(len([]rune(strings.ToLower(str)))-subStrLen) {
+			i64s[i] = 0
+			continue
+		} else if subStrLen == 0 {
+			i64s[i] = pos + 1
+			continue
+		}
+		slice := string([]rune(strings.ToLower(str))[pos:])
+		idx := strings.Index(slice, strings.ToLower(subStr))
```
Comment (fzhedu) [inline expression/builtin_string_vec.go]: If `subStr` is constant, it is better to convert the `subStr` to the lower case when copying it to a column, then avoiding invocking `ToLower` here.

Comment (b41sh) [reply expression/builtin_string_vec.go]: ok

File: expression/builtin_string_vec.go [hunk 5]
```diff
@@ -333,11 +333,64 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
...
+	}
+	// store positions in result
+	if err := b.args[2].VecEvalInt(b.ctx, input, result); err != nil {
+		return err
+	}
+
+	result.MergeNulls(buf, buf1)
+	i64s := result.Int64s()
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
```
Comment (SunRunAway) [inline expression/builtin_string_vec.go]: Consider this case, b.args[2] is null but the result is not null.
```mysql
mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
+----------------------------------+
| LOCATE('bar', 'foobarbar', NULL) |
+----------------------------------+
|                                0 |
+----------------------------------+
1 row in set (0.00 sec)
```

File: expression/builtin_string_vec.go [hunk 6]
```diff
@@ -333,11 +333,64 @@ func (b *builtinConcatSig) vecEvalString(input *chunk.Chunk, result *chunk.Colum
...
+			continue
+		}
+		subStr := strings.ToLower(buf.GetString(i))
+		str := strings.ToLower(buf1.GetString(i))
```
Comment (SunRunAway) [inline expression/builtin_string_vec.go]: we do not need to call `strings.ToLower` until `strings.Index`, how about:
```go
subStr := buf.GetSring(i)
str := buf1.GetString(i)
...
...
...
slice := string([]rune(str)[pos:])
slice, subStr := strings.ToLower(slice), strings.ToLower(subStr)
idx := strings.Index(slice, subStr)
```

Comment (b41sh) [reply expression/builtin_string_vec.go]: ok

File: expression/builtin_string_vec.go (hunk 5 above)
Comment (b41sh) [reply expression/builtin_string_vec.go]: My test result is correct

```sql
mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
+----------------------------------+
| LOCATE('bar', 'foobarbar', NULL) |
+----------------------------------+
|                             NULL |
+----------------------------------+
1 row in set (0.00 sec)
```

Comment (fzhedu) [reply expression/builtin_string_vec.go]: can args[0] and args[1] be null?

Comment (b41sh) [reply expression/builtin_string_vec.go]: yes, any args is null will result in null

Comment (qw4990) [reply expression/builtin_string_vec.go]: > Consider this case, b.args[2] is null but the result is not null.
> 
> ```sql
> mysql> SELECT LOCATE('bar', 'foobarbar', NULL);
> +----------------------------------+
> | LOCATE('bar', 'foobarbar', NULL) |
> +----------------------------------+
> |                                0 |
> +----------------------------------+
> 1 row in set (0.00 sec)
> ```

In MySQL 8.x, this SQL returns `NULL`.

**Pull request merged**